File: weird_captcha_gym/shared_runtime/server/incubator_graders/curio_tray.py
from __future__ import annotations

from typing import Any


MECHANIC_ID = "curio_tray"
TRAY_CAPACITY = 7


def _fail(feedback: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": feedback}


def _overlap(first: dict[str, Any], second: dict[str, Any]) -> bool:
    return (
        float(first["x"]) < float(second["x"]) + float(second["width"])
        and float(second["x"]) < float(first["x"]) + float(first["width"])
        and float(first["y"]) < float(second["y"]) + float(second["height"])
        and float(second["y"]) < float(first["y"]) + float(first["height"])
    )


def _accessible(items: list[dict[str, Any]], remaining: set[str]) -> list[str]:
    visible: list[str] = []
    for item in items:
        item_id = str(item.get("id") or "")
        if item_id not in remaining:
            continue
        covered = any(
            other is not item
            and str(other.get("id") or "") in remaining
            and int(other.get("z")) > int(item.get("z"))
            and _overlap(item, other)
            for other in items
        )
        if not covered:
            visible.append(item_id)
    return visible


def _bind(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> str | None:
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID:
        return "payload mechanic mismatch"
    if str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return "ground-truth mechanic mismatch"
    if str(public_state.get("mechanic_id") or "") != MECHANIC_ID:
        return "public-state mechanic mismatch"
    challenge_id = str(ground_truth.get("challenge_id") or "")
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id:
        return "stale challenge"
    if str(public_state.get("challenge_id") or "") != challenge_id:
        return "public-state challenge mismatch"
    task_id = str(ground_truth.get("task_id") or "")
    if not task_id or str(payload.get("task_id") or "") != task_id:
        return "payload task mismatch"
    if str(public_state.get("task_id") or "") != task_id:
        return "public-state task mismatch"
    return None


def _contract(ground_truth: dict[str, Any], public_state: dict[str, Any]) -> tuple[list[dict[str, Any]], int]:
    hidden = ground_truth.get("items")
    public = (public_state.get("cabinet") or {}).get("items")
    if not isinstance(hidden, list) or not hidden or not isinstance(public, list) or len(public) != len(hidden):
        raise ValueError("item contract is malformed")
    public_by_id = {str(item.get("id") or ""): item for item in public if isinstance(item, dict)}
    if len(public_by_id) != len(public):
        raise ValueError("public item ids are not unique")
    projected: list[dict[str, Any]] = []
    for item in hidden:
        if not isinstance(item, dict) or not str(item.get("id") or ""):
            raise ValueError("hidden item is malformed")
        item_id = str(item["id"])
        public_item = public_by_id.get(item_id)
        if public_item is None:
            raise ValueError("public item set differs from hidden item set")
        for key in ("type", "name", "icon", "tone", "x", "y", "width", "height", "z"):
            if public_item.get(key) != item.get(key):
                raise ValueError(f"public item geometry differs for {item_id}")
        if not isinstance(item.get("type_index"), int):
            raise ValueError("hidden type index is malformed")
        projected.append(item)
    type_count = ground_truth.get("type_count")
    if isinstance(type_count, bool) or not isinstance(type_count, int) or not 3 <= type_count <= 9:
        raise ValueError("type-count contract is malformed")
    counts = [0] * type_count
    for item in projected:
        type_index = int(item["type_index"])
        if not 0 <= type_index < type_count:
            raise ValueError("item type lies outside the type contract")
        counts[type_index] += 1
    if counts != [3] * type_count:
        raise ValueError("every curio type must have exactly three copies")
    if public_state.get("tray_capacity") != TRAY_CAPACITY:
        raise ValueError("tray capacity differs from hidden contract")
    return projected, type_count


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    binding_error = _bind(payload, ground_truth, public_state)
    if binding_error:
        return _fail(binding_error)
    if ground_truth.get("control_condition") != public_state.get("control_condition"):
        return _fail("public interaction condition differs from Curio Tray contract")
    condition = ground_truth.get("control_condition") or {}
    interaction = str(condition.get("interaction") or "")
    expected_source = {"simplified": "proxy_pick", "full": "pile_click"}.get(interaction)
    if expected_source is None:
        return _fail("Curio Tray interaction condition is invalid")
    try:
        items, type_count = _contract(ground_truth, public_state)
    except (TypeError, ValueError) as exc:
        return _fail(f"invalid Curio Tray contract: {exc}")

    item_by_id = {str(item["id"]): item for item in items}
    remaining = set(item_by_id)
    tray: list[str] = []
    transcript = payload.get("picks")
    if not isinstance(transcript, list) or len(transcript) > len(items):
        return _fail("pick transcript is missing or too long")
    triple_count = 0
    failed = False
    for sequence, event in enumerate(transcript, start=1):
        if failed:
            return _fail("transcript continues after a terminal tray overflow")
        if not isinstance(event, dict) or event.get("sequence") != sequence:
            return _fail(f"pick {sequence} has an invalid sequence")
        if event.get("input_source") != expected_source:
            return _fail(f"pick {sequence} uses the wrong interaction input")
        item_id = str(event.get("item_id") or "")
        if item_id not in remaining:
            return _fail(f"pick {sequence} repeats or names an unknown curio")
        if item_id not in set(_accessible(items, remaining)):
            return _fail(f"pick {sequence} selected an occluded curio")
        item_type = str(item_by_id[item_id]["type"])
        remaining.remove(item_id)
        tray.append(item_type)
        cleared_type: str | None = None
        outcome = "pick"
        if tray.count(item_type) == 3:
            tray = [value for value in tray if value != item_type]
            cleared_type = item_type
            outcome = "triple_clear"
            triple_count += 1
        if len(tray) >= TRAY_CAPACITY and remaining:
            failed = True
        expected = {
            "outcome": outcome,
            "cleared_type": cleared_type,
            "tray_after": tray,
            "tray_size_after": len(tray),
            "remaining_count": len(remaining),
        }
        for field, value in expected.items():
            if event.get(field) != value:
                return _fail(f"pick {sequence} has inconsistent {field}: expected {value!r}")
        if failed:
            break

    if payload.get("remaining_ids") != sorted(remaining):
        return _fail("submitted remaining curios do not match replay")
    if payload.get("tray") != tray:
        return _fail("submitted tray does not match replay")
    if payload.get("pick_count") != len(transcript):
        return _fail("submitted pick count does not match replay")
    if payload.get("triple_count") != triple_count:
        return _fail("submitted triple count does not match replay")
    completed = payload.get("completed") is True
    passed = completed and not failed and not remaining and not tray and triple_count == type_count
    return {
        "graded": True,
        "passed": passed,
        "feedback": (
            f"curio replay: {len(items) - len(remaining)}/{len(items)} picked; "
            f"{triple_count}/{type_count} triples cleared; tray {len(tray)}/{TRAY_CAPACITY}; "
            f"{'overflow' if failed else 'empty cabinet' if passed else 'incomplete'}"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    return {
        "solution_order": ground_truth.get("solution_order") or [],
        "instruction": "Click the exposed curios in the supplied solution order, then appraise the empty tray.",
    }

File: weird_captcha_gym/shared_runtime/app/mechanics/curio_tray.js
(() => {
  "use strict";

  const CAPACITY = 7;
  const model = {
    state: null,
    items: [],
    remaining: new Set(),
    tray: [],
    picks: [],
    triples: 0,
    busy: false,
    terminal: false,
    ready: false,
    interaction: "full",
    helpers: null,
  };

  const clean = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");

  const overlap = (first, second) => (
    Number(first.x) < Number(second.x) + Number(second.width)
    && Number(second.x) < Number(first.x) + Number(first.width)
    && Number(first.y) < Number(second.y) + Number(second.height)
    && Number(second.y) < Number(first.y) + Number(first.height)
  );

  function accessibleIds() {
    return model.items.filter((item) => model.remaining.has(item.id) && !model.items.some((other) => (
      other.id !== item.id
      && model.remaining.has(other.id)
      && Number(other.z) > Number(item.z)
      && overlap(item, other)
    ))).map((item) => item.id);
  }

  function itemById(itemId) {
    return model.items.find((item) => item.id === itemId) || null;
  }

  function iconMarkup(item) {
    return `<span class="curio-icon curio-tone-${clean(item.tone)}" aria-hidden="true">${clean(item.icon)}</span>`;
  }

  function pileMarkup() {
    const exposed = new Set(accessibleIds());
    return model.items.filter((item) => model.remaining.has(item.id)).map((item) => {
      const accessible = exposed.has(item.id);
      const classes = `curio-item curio-tone-${clean(item.tone)}${accessible ? " is-exposed" : " is-occluded"}`;
      return `<button type="button" class="${classes}" data-item-id="${clean(item.id)}" data-accessible="${accessible ? "true" : "false"}" style="left:${Number(item.x)}px;top:${Number(item.y)}px;width:${Number(item.width)}px;height:${Number(item.height)}px;z-index:${Number(item.z)}" aria-label="${clean(item.name)}${accessible ? " exposed" : " occluded"}">
        <span class="curio-rim"></span>${iconMarkup(item)}<span class="curio-name">${clean(item.name)}</span><span class="curio-depth">${accessible ? "EXPOSED" : "UNDER GLASS"}</span>
      </button>`;
    }).join("");
  }

  function proxyMarkup() {
    const visible = accessibleIds();
    if (!visible.length) return '<p class="proxy-empty">NO EXPOSED CURIOS</p>';
    return visible.map((itemId, index) => {
      const item = itemById(itemId);
      return `<button type="button" class="curio-proxy" data-item-id="${clean(itemId)}" aria-label="Pick exposed ${clean(item.name)}"><b>${String(index + 1).padStart(2, "0")}</b>${iconMarkup(item)}<span><strong>${clean(item.name)}</strong><small>EXPOSED PICK</small></span></button>`;
    }).join("");
  }

  function trayMarkup() {
    const slots = [];
    for (let index = 0; index < CAPACITY; index += 1) {
      const item = model.tray[index];
      slots.push(item
        ? `<div class="tray-slot is-filled curio-tone-${clean(item.tone)}" data-slot-index="${index}">${iconMarkup(item)}<small>${clean(item.name)}</small></div>`
        : `<div class="tray-slot" data-slot-index="${index}"><span>${String(index + 1).padStart(2, "0")}</span></div>`);
    }
    return slots.join("");
  }

  function updatePanels() {
    const exposed = document.getElementById("curio-exposed-count");
    const remaining = document.getElementById("curio-remaining-count");
    const trayCount = document.getElementById("curio-tray-count");
    const tripleCount = document.getElementById("curio-triple-count");
    const tray = document.getElementById("curio-tray");
    const proxy = document.getElementById("curio-proxy-list");
    const submit = document.getElementById("curio-submit");
    if (exposed) exposed.textContent = String(accessibleIds().length).padStart(2, "0");
    if (remaining) remaining.textContent = String(model.remaining.size).padStart(2, "0");
    if (trayCount) trayCount.textContent = `${model.tray.length} / ${CAPACITY}`;
    if (tripleCount) tripleCount.textContent = String(model.triples).padStart(2, "0");
    if (tray) tray.innerHTML = trayMarkup();
    if (proxy) proxy.innerHTML = proxyMarkup();
    if (submit) submit.disabled = model.busy || model.terminal;
  }

  function flash(message, status = "idle") {
    model.helpers.setReadout(message, status);
    const shell = document.querySelector(".curio-tray-captcha");
    shell?.classList.remove("is-jolt");
    void shell?.offsetWidth;
    shell?.classList.add("is-jolt");
  }

  function clearFailure() {
    document.querySelector(".curio-verdict-fail")?.remove();
    document.querySelector(".curio-tray-captcha")?.classList.remove("is-fresh-fail");
  }

  function bindPickButtons() {
    document.querySelectorAll(".curio-item.is-exposed").forEach((button) => {
      button.addEventListener("click", () => {
        if (model.interaction === "full") attemptPick(button.dataset.itemId, "pile_click");
      });
    });
    document.querySelectorAll(".curio-proxy").forEach((button) => {
      button.addEventListener("click", () => {
        if (model.interaction === "simplified") attemptPick(button.dataset.itemId, "proxy_pick");
      });
    });
  }

  async function submit(completed) {
    if (!model.state || model.busy && !model.terminal) return;
    model.busy = true;
    document.querySelectorAll("button").forEach((button) => { button.disabled = true; });
    model.helpers.setReadout(completed ? "APPRAISING EMPTY TRAY…" : "VOIDING ATTEMPT…", "pending");
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      picks: model.picks,
      tray: model.tray.map((item) => item.type),
      remaining_ids: Array.from(model.remaining).sort(),
      pick_count: model.picks.length,
      triple_count: model.triples,
      completed,
    };
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify(payload),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        const shell = document.querySelector(".curio-tray-captcha");
        shell?.classList.add("is-pass");
        shell?.insertAdjacentHTML("beforeend", '<div class="curio-verdict curio-verdict-pass"><small>CABINET INVENTORY RECONCILED</small><strong>PASS</strong></div>');
        model.helpers.setReadout("PASS · EMPTY TRAY CERTIFIED", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await model.helpers.render(outcome.state);
        const shell = document.querySelector(".curio-tray-captcha");
        shell?.classList.add("is-fresh-fail");
        shell?.insertAdjacentHTML("beforeend", '<div class="curio-verdict curio-verdict-fail"><small>ATTEMPT VOID · NEW DRAW</small><strong>FAIL</strong></div>');
        model.helpers.setReadout("FAIL · NEW CABINET DRAWN", "error");
        window.setTimeout(() => document.querySelector(".curio-verdict-fail")?.remove(), 1800);
      } else {
        model.busy = false;
        model.terminal = false;
        model.helpers.setReadout("FAIL · TRAY NOT CERTIFIED", "error");
        document.querySelectorAll("button").forEach((button) => { button.disabled = false; });
      }
    } catch (_error) {
      model.busy = false;
      model.terminal = false;
      model.helpers.setReadout("FAIL · REGISTER OFFLINE", "error");
      document.querySelectorAll("button").forEach((button) => { button.disabled = false; });
    }
  }

  function overflowFailure() {
    model.terminal = true;
    const shell = document.querySelector(".curio-tray-captcha");
    shell?.classList.add("is-fresh-fail");
    shell?.insertAdjacentHTML("beforeend", '<div class="curio-verdict curio-verdict-fail"><small>SEVEN UNMATCHED CURIOS</small><strong>VOID</strong></div>');
    flash("TRAY FULL · CABINET DRAW VOID", "error");
    submit(false);
  }

  function attemptPick(itemId, inputSource) {
    if (model.busy || model.terminal) return;
    clearFailure();
    const exposed = accessibleIds();
    if (!exposed.includes(itemId)) {
      flash("THAT CURIO IS UNDER ANOTHER", "error");
      return;
    }
    const item = itemById(itemId);
    if (!item || !model.remaining.has(itemId)) return;
    model.remaining.delete(itemId);
    model.tray.push(item);
    let outcome = "pick";
    let clearedType = null;
    if (model.tray.filter((entry) => entry.type === item.type).length === 3) {
      model.tray = model.tray.filter((entry) => entry.type !== item.type);
      model.triples += 1;
      outcome = "triple_clear";
      clearedType = item.type;
    }
    const event = {
      sequence: model.picks.length + 1,
      item_id: itemId,
      input_source: inputSource,
      outcome,
      cleared_type: clearedType,
      tray_after: model.tray.map((entry) => entry.type),
      tray_size_after: model.tray.length,
      remaining_count: model.remaining.size,
    };
    model.picks.push(event);
    renderBoard();
    updatePanels();
    bindPickButtons();
    if (model.tray.length >= CAPACITY && model.remaining.size) {
      overflowFailure();
      return;
    }
    if (!model.remaining.size) {
      model.ready = model.tray.length === 0;
      document.querySelector(".curio-tray-captcha")?.classList.add("is-ready");
      flash("CABINET CLEAR · APPRAISE THE EMPTY TRAY", "passed");
    } else if (outcome === "triple_clear") {
      flash(`TRIPLE CLEAR · ${item.name} VANISHED`, "idle");
    } else {
      flash(`${item.name} STORED · ${model.tray.length} / ${CAPACITY} SLOTS`, "idle");
    }
  }

  function renderBoard() {
    const pile = document.getElementById("curio-pile");
    if (pile) pile.innerHTML = pileMarkup();
  }

  function resetAttempt() {
    if (model.busy) return;
    model.remaining = new Set(model.items.map((item) => item.id));
    model.tray = [];
    model.picks = [];
    model.triples = 0;
    model.terminal = false;
    model.ready = false;
    document.querySelector(".curio-tray-captcha")?.classList.remove("is-pass", "is-ready", "is-fresh-fail");
    document.querySelectorAll(".curio-verdict").forEach((node) => node.remove());
    renderBoard();
    updatePanels();
    bindPickButtons();
    flash("DRAW RESET · SELECT AN EXPOSED CURIO", "idle");
  }

  async function render(state, helpers) {
    document.body.dataset.mechanic = "curio-tray";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    const cabinet = state.cabinet || {};
    Object.assign(model, {
      state,
      items: (cabinet.items || []).map((item) => ({...item, id: String(item.id)})),
      remaining: new Set((cabinet.items || []).map((item) => String(item.id))),
      tray: [],
      picks: [],
      triples: 0,
      busy: false,
      terminal: false,
      ready: false,
      interaction: state.control_condition?.interaction || "full",
      helpers,
    });
    const interaction = model.interaction;
    helpers.app.innerHTML = `
      <section class="curio-tray-captcha" data-interaction="${clean(interaction)}" data-challenge-id="${clean(state.challenge_id)}">
        <header class="curio-head">
          <div class="curio-brand"><span>THE CURIO CABINET · INTAKE DESK 07</span><h1>${clean(state.prompt)}</h1></div>
          <div class="curio-ticket"><small>TRAY CAPACITY</small><strong>07</strong><i>UNMATCHED SLOTS</i></div>
        </header>
        <main class="curio-workbench">
          <section class="curio-cabinet-panel">
            <div class="curio-panel-label"><span>GLASS CABINET / LAYERED DRAW</span><b>${Number(cabinet.stack_count || 0)} STACKS</b></div>
            <div class="curio-cabinet" id="curio-cabinet" style="--cabinet-h:${Number(cabinet.height || 455)}px"><div class="curio-pile" id="curio-pile">${pileMarkup()}</div><div class="cabinet-glare"></div></div>
            <div class="curio-cabinet-foot"><span><i class="legend-dot exposed"></i> EXPOSED</span><span><i class="legend-dot occluded"></i> UNDER ANOTHER CURIO</span><b>CLICK ONLY WHAT IS ON TOP</b></div>
          </section>
          <aside class="curio-side-panel">
            <div class="curio-stat-row"><span>EXPOSED NOW</span><b id="curio-exposed-count">00</b></div>
            <div class="curio-stat-row"><span>LEFT IN CABINET</span><b id="curio-remaining-count">00</b></div>
            <div class="curio-stat-row"><span>TRIPLES CLEARED</span><b id="curio-triple-count">00</b></div>
            <section class="curio-proxy-wrap" data-proxy="${interaction === "simplified" ? "true" : "false"}">
              <header><span>${interaction === "simplified" ? "EXPOSED PICKS / PROXY" : "EXPOSED PICKS / MAP"}</span><small>${interaction === "simplified" ? "CLICK A ROW TO LIFT IT" : "USE THE CABINET"}</small></header>
              <div class="curio-proxy-list" id="curio-proxy-list">${proxyMarkup()}</div>
            </section>
            <section class="curio-rule-card"><b>TRIPLE RULE</b><p>Three matching curios disappear together and free their slots.</p><b>LOSS RULE</b><p>Seven unmatched curios fill the tray and void the draw.</p></section>
          </aside>
          <section class="curio-tray-panel">
            <div class="curio-panel-label"><span>THE SEVEN-SLOT TRAY</span><b id="curio-tray-count">0 / 7</b></div>
            <div class="curio-tray" id="curio-tray">${trayMarkup()}</div>
            <div class="curio-tray-note"><span>Matching sets clear automatically.</span><b>KEEP A SLOT OPEN FOR THE THIRD</b></div>
          </section>
        </main>
        <footer class="curio-foot">
          <button type="button" class="curio-reset" id="curio-reset">↺ RESET DRAW</button>
          <div class="readout" data-status="idle">SELECT AN EXPOSED CURIO TO BEGIN</div>
          <button type="button" class="curio-submit" id="curio-submit">${clean(state.submit_label || "APPRAISE EMPTY TRAY")}</button>
        </footer>
        ${helpers.cheatPanelTemplate()}
      </section>`;
    updatePanels();
    bindPickButtons();
    document.getElementById("curio-reset")?.addEventListener("click", resetAttempt);
    document.getElementById("curio-submit")?.addEventListener("click", () => submit(model.ready));
    helpers.installCheatPanel();
    window.curioTrayModel = model;
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.curio_tray = {rootSelector: ".curio-tray-captcha", render};
})();
